keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,546 @@
+
+There should only be two such cases, and in normal operation, only one would happen frequently:
+
+- The first, and most clear case of a container changing state without the Kubelet triggering that state change is when a container stops. Containers can exit gracefully, or be OOM killed, and the Kubelet would not know.
+   - We will also introduce events when the container is created as well as is started. This will help us reduce the relisting that takes placed while the kubelet waits for the container to start.
+   - Although kubelet initiates the container deletion, for sake of increased validation we are also introducing the event to denote that from the runtime.
+- The second, and less likely case is when another entity comes and changes the state of the node.
+	- For container related events (such as a container creating, starting, stopping or being killed), this can appear as a user calling [crictl](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/crictl.md) manually, or even using the runtime directly.
+
+The Kubelet currently covers each of thse cases quite easily: by listing all of the resources on the node, it will have an accurate picture after the amount of time of its [poll interval](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/kubelet.go#L162). For each of these cases, a new CRI-based events API can be made, using [gRPC server streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc). This way, the entity closest to the activity of the containers and pods (the CRI implementation) can be responsible for informing the Kubelet of their behavior directly.
+
+### User Stories
+
+- As a cluster administrator I want to enable `Evented PLEG` feature of the kubelet for better performance with as little infrastructure overhead as possible.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- PLEG is very core to the container status handling in the kubelet. Hence any miscalculation there would result in unpredictable behaviour not just for the node but for an entire cluster.
+  - To reduce the risk of regression, this feature initially will be available only as an opt-in.
+  - Users can disable this feature to make kubelet use existing relisting based PLEG.
+- Another risk is the CRI implementation could have a buggy event emitting system, and miss pod lifecycle events.
+  - A mitigation is a `kube_pod_missed_events` metric, which the Kubelet could report when a lifecycle event is registered that wasn't triggered by an event, but rather by changes of state between lists.
```

Comment:
we may want to refine this metric name during implementation, but i agree with concept.

Reply:
we have a metric today that is `pleg_discard_events` for awareness, so it may be worth a pass to see if the name will confuse with existing metrics or not.